File: src/476.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def findComplement(self, num: int) -> int:
        retVal = num

        i = 31
        while retVal & (1 << i) == 0:
            i -= 1

        while i >= 0:
            retVal ^= (1 << i)
            i -= 1

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for num in [5, 1]:
            # /* Example
            #  *  Input: num = 5
            #  *  Output: 2
            #  *
            #  *  Input: num = 1
            #  *  Output: 0
            #  */
            logging.info("Input: num = %s", num)

            retVal = pSolution.findComplement(num)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/476.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int findComplement(int num) {
        int retVal = num;

        int i = 31;
        while ((retVal & ((uint32_t)1 << i)) == 0) {
            i--;
        }

        while (i >= 0) {
            retVal ^= (1 << i--);
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        int num;
    };
    vector<subject> testData{{5}, {1}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: num = 5
     *  Output: 2
     *
     *  Input: num = 1
     *  Output: 0
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: num = " << testData[i].num << "\n";

        answer = cSolution.findComplement(testData[i].num);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
